### PR TITLE
feat: bump badger to v4

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary

badger v2 is ancient

use a separate folder for badgerv4 DB and drop old data

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## Related issues

Closes https://github.com/elastic/apm-server/issues/11546

### CODE REVIEW

File: x-pack/apm-server/sampling/processor_test.go [hunk 1]
```diff
@@ -589,6 +589,7 @@ func TestStorageMonitoring(t *testing.T) {
 }
 
 func TestStorageGC(t *testing.T) {
+	t.Skip("SKIP")
```
Comment (marclop) [inline x-pack/apm-server/sampling/processor_test.go:592]: Is there a compelling reason to sneak this into this commit?

File: x-pack/apm-server/main.go [hunk 2]
```diff
@@ -117,6 +119,23 @@ func newTailSamplingProcessor(args beater.ServerParams) (*sampling.Processor, er
 		return nil, fmt.Errorf("failed to create Elasticsearch client for tail-sampling: %w", err)
 	}
 
+	if entries, err := os.ReadDir(oldTailSamplingStorageDir); err == nil {
```
Comment (marclop) [inline x-pack/apm-server/main.go:122]: I don't see any tests ensuring this works, could you add some, please?

**Review by marclop — COMMENTED** [review]
Can you add some tests for the new behavior?

File: x-pack/apm-server/main.go [hunk 3]
```diff
@@ -117,6 +119,24 @@ func newTailSamplingProcessor(args beater.ServerParams) (*sampling.Processor, er
 		return nil, fmt.Errorf("failed to create Elasticsearch client for tail-sampling: %w", err)
 	}
 
+	oldStorageDir := paths.Resolve(paths.Data, oldTailSamplingStorageDir)
+	if entries, err := os.ReadDir(oldStorageDir); err == nil {
```
Comment (1pkg) [inline x-pack/apm-server/main.go:123]: Do we want to log the error if the syscall to open a directory has failed here? i.e. specifically checking for `err != nil && err != io.EOF`

**Review by 1pkg — APPROVED** [review]

**Review by marclop — CHANGES_REQUESTED** [review]
Can you add a changelog entry? This seems like the kind of change that needs a changelog entry.

**Comment by simitt** [comment]
@kruskall I think this should be closed based on the findings and the workaround you applied.

**Pull request closed without merging**